File: python_conversion/matrix_3d.py
"""
Matrix3D class for 3x3 transformation matrices.
"""

from xyz import XYZ

class Matrix3D:
    """
    3x3 transformation matrix. Stored as three rows, each an XYZ vector.

    This class supports:
      - Construction from rows or columns
      - Transposition
      - Matrix-vector multiplication
      - Matrix-matrix multiplication
      - Construction of orthogonal transformation matrices from vector pairs or triples
    """

    def __init__(self, row1, row2, row3):
        """
        Initialize a Matrix3D from three row vectors.
        Args:
            row1 (XYZ): First row.
            row2 (XYZ): Second row.
            row3 (XYZ): Third row.
        """
        self.row1 = row1
        self.row2 = row2
        self.row3 = row3

    @staticmethod
    def from_columns(col1, col2, col3):
        """
        Construct a Matrix3D from three column vectors (XYZ).
        Internally, the matrix is stored by rows, so this method transposes the columns.
        Args:
            col1 (XYZ): First column.
            col2 (XYZ): Second column.
            col3 (XYZ): Third column.
        Returns:
            Matrix3D: The resulting matrix.
        """
        return Matrix3D.transpose(Matrix3D(col1, col2, col3))

    def __str__(self):
        """
        String representation of the matrix.
        Returns:
            str: String showing all three rows.
        """
        return f"Matrix with rows {self.row1} {self.row2} {self.row3}"

    @staticmethod
    def identity():
        """
        Return the 3x3 identity matrix.
        Returns:
            Matrix3D: The identity matrix.
        """
        return Matrix3D(XYZ(1, 0, 0), XYZ(0, 1, 0), XYZ(0, 0, 1))

    @staticmethod
    def transpose(m):
        """
        Return the transpose of matrix m.
        Args:
            m (Matrix3D): The matrix to transpose.
        Returns:
            Matrix3D: The transposed matrix.
        """
        return Matrix3D(
            XYZ(m.row1.x, m.row2.x, m.row3.x),
            XYZ(m.row1.y, m.row2.y, m.row3.y),
            XYZ(m.row1.z, m.row2.z, m.row3.z)
        )

    @staticmethod
    def mult(m, v):
        """
        Multiply matrix m by vector v (XYZ).
        Args:
            m (Matrix3D): The matrix.
            v (XYZ): The vector.
        Returns:
            XYZ: The resulting vector.
        """
        return XYZ(
            XYZ.dot(m.row1, v),
            XYZ.dot(m.row2, v),
            XYZ.dot(m.row3, v)
        )

    @staticmethod
    def product(m1, m2):
        """
        Multiply two matrices (m1 * m2).
        Args:
            m1 (Matrix3D): First matrix.
            m2 (Matrix3D): Second matrix.
        Returns:
            Matrix3D: The product matrix.
        """
        m2t = Matrix3D.transpose(m2)
        return Matrix3D.from_columns(
            Matrix3D.mult(m1, m2t.row1),
            Matrix3D.mult(m1, m2t.row2),
            Matrix3D.mult(m1, m2t.row3)
        )

    @staticmethod
    def from_pairs(t1, u1, t2, u2):
        """
        Construct an orthogonal transformation matrix M such that:
        M * t1 = t2 and M * u1 = u2.
        Assumes t's and u's are unit vectors (not colinear) with the same dot product.
        Args:
            t1, u1 (XYZ): Source vectors.
            t2, u2 (XYZ): Target vectors.
        Returns:
            Matrix3D: The resulting transformation matrix.
        """
        v1 = XYZ.unit(XYZ.cross(t1, u1))
        v2 = XYZ.unit(XYZ.cross(t2, u2))
        w1 = XYZ.cross(v1, u1)
        w2 = XYZ.cross(v2, u2)
        return Matrix3D.from_orthog3(u2, v2, w2, u1, v1, w1)

    @staticmethod
    def from_orthog3(u1, v1, w1, u2, v2, w2):
        """
        Construct an orthogonal transformation matrix M such that:
        M * u1 = u2, M * v1 = v2, M * w1 = w2.
        Assumes each (u, v, w) triple is an orthonormal set.
        If M1 has columns u1, v1, w1 and M2 has columns u2, v2, w2,
        then the answer is M2 * M1^T.
        Args:
            u1, v1, w1 (XYZ): Source orthonormal set.
            u2, v2, w2 (XYZ): Target orthonormal set.
        Returns:
            Matrix3D: The resulting transformation matrix.
        """
        return Matrix3D.product(
            Matrix3D.from_columns(u2, v2, w2),
            Matrix3D(u1, v1, w1)
        )

File: original_java_code/GrowthForm4/src/Matrix3D.java


// 3x3 transformation matrix.  Stored as three rows, each an XYZ

class Matrix3D implements java.io.Serializable {
  private static final long serialVersionUID = 1L;
  XYZ row1;
  XYZ row2;
  XYZ row3;

  public Matrix3D(XYZ newrow1, XYZ newrow2, XYZ newrow3){  // constructor from rows
    row1 = newrow1;
    row2 = newrow2;
    row3 = newrow3;
  }

  public static Matrix3D fromColumns(XYZ col1, XYZ col2, XYZ col3){ // construct from columns
    return transpose(new Matrix3D(col1, col2, col3));
  }

  public String toString(){
    return "Matrix with rows " + this.row1 + this.row2 + this.row3;
  }

  public static Matrix3D identity(){
    return new Matrix3D(new XYZ(1,0,0),
                        new XYZ(0,1,0),
                        new XYZ(0,0,1));
  }

  public static Matrix3D transpose(Matrix3D m){
    return new Matrix3D(new XYZ(m.row1.x, m.row2.x, m.row3.x),
                        new XYZ(m.row1.y, m.row2.y, m.row3.y),
                        new XYZ(m.row1.z, m.row2.z, m.row3.z));
  }

  public static XYZ mult(Matrix3D m, XYZ v) {  // matrix-vector multiplication
    return new XYZ(XYZ.dot(m.row1,v), XYZ.dot(m.row2,v), XYZ.dot(m.row3,v));
  }

  public static Matrix3D product(Matrix3D m1, Matrix3D m2){   // matrix-matrix prod
    Matrix3D m2t = transpose(m2);     // so row of m2t is really column of m2
    return fromColumns(mult(m1,m2t.row1), mult(m1,m2t.row2), mult(m1,m2t.row3));
  }

  // the orthog transformation matrix M such that M t1 = t2 and M u1 = u2
  // assumes t's and u's are unit vectors (not colinear) with same dot prod.
  public static Matrix3D fromPairs(XYZ t1, XYZ u1, XYZ t2, XYZ u2){
    XYZ v1 = XYZ.unit(XYZ.cross(t1,u1));
    XYZ v2 = XYZ.unit(XYZ.cross(t2,u2));
    XYZ w1 = XYZ.cross(v1,u1);
    XYZ w2 = XYZ.cross(v2,u2);
    return fromOrthog3(u2,v2,w2, u1,v1,w1);
  }

  // the orthog transformation matrix M such that M u1 = u2 and M v1 = v2
  // and M w1 = w2.  Assumes each uvw triple is an orthonormal set.
  // If M1 has as columns u1v1w1 and M2 has columns u2v2w2,  Ans = M2 M1^T
  public static Matrix3D fromOrthog3(XYZ u1, XYZ v1, XYZ w1,
                                     XYZ u2, XYZ v2, XYZ w2){
    return product(fromColumns(u2,v2,w2), new Matrix3D(u1,v1,w1));
  }
}